File: movie-infrastructures/storage/src/main/java/com/movie/storage/movie/repository/ReservationJpaRepository.java
package com.movie.storage.movie.repository;

import com.movie.storage.movie.entity.ReservationEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;

public interface ReservationJpaRepository extends JpaRepository<ReservationEntity, Long> {

    List<ReservationEntity> findById_ScheduleIdAndId_SeatIdIn(Long id_ScheduleId, List<Long> id_SeatId);

    @Modifying
    @Query("update ReservationEntity r set r.userId = :userId, r.reservedAt = :reservedAt where r.id.scheduleId = :scheduleId and r.id.seatId in :seatIds")
    void updateAsReserved(@Param("scheduleId") Long scheduleId,
                          @Param("seatIds") List<Long> seatIds,
                          @Param("userId") Long userId,
                          @Param("reservedAt") LocalDateTime reservedAt);
}

File: movie/infra/src/main/kotlin/com/example/movie/persistence/reservation/repository/ReservationJpaRepository.kt
package com.example.movie.persistence.reservation.repository

import com.example.movie.persistence.reservation.entity.ReservationEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface ReservationJpaRepository : JpaRepository<ReservationEntity, Long> {
    @Query("SELECT r FROM ReservationEntity r WHERE r.screening.id = :screeningId AND r.seat.id IN :seatIds")
    fun findByScreeningIdAndSeatIds(
        @Param("screeningId") screeningId: Long,
        @Param("seatIds") seatIds: List<Long>
    ): List<ReservationEntity>

    fun findByScreeningIdAndUserId(screeningId: Long, userId: Long): List<ReservationEntity>
}